crates/iceberg-catalog/src/service/task_queue/mod.rs

```diff
@@ -1,132 +1,219 @@
+    async fn run(self, poll_interval: Duration) {
+        loop {
+            for (queue_name, queue_tx) in &self.queues {
+                match self.queue.pick_new_task(queue_name).await {
+                    Ok(Some(task)) => {
+                        if let Err(e) = tryhard::retry_fn(|| queue_tx.send(task.clone()))
+                            .retries(5)
+                            .fixed_backoff(Duration::from_millis(100))
+                            .await
+                        {
+                            // TODO: Do we even want to record_failure for this?
+                            self.queue
+                                .retrying_record_failure(
+                                    &task,
+                                    &format!("Failed to forward task to task queue handler {e}"),
+                                )
+                                .await;
+                        };
+                    }
+                    Ok(None) => {
+                        tracing::trace!("No task available for queue '{queue_name}'");
+                        // No task available
+                    }
+                    Err(e) => {
+                        tracing::error!("Failed to pick new task: '{}'", e.error);
+                    }
+                }
+            }
+            tokio::time::sleep(poll_interval).await;
+        }
     }
 
-    #[tracing::instrument(skip(self))]
-    pub(crate) async fn queue_tabular_purge(
-        &self,
-        task: TabularPurgeInput,
-    ) -> crate::api::Result<()> {
-        self.tabular_purge.enqueue(task).await
+    async fn outer_run(mut self, poll_interval: Duration) -> anyhow::Result<()> {
+        let mut queue_tasks = vec![];
+        let mut qs = HashMap::with_capacity(0);
+        std::mem::swap(&mut self.registered_queues, &mut qs);
+        for (
+            name,
+            RegisteredQueue {
+                config,
+                queue_task: task_fn,
+            },
+        ) in qs
+        {
+            tracing::info!(
+                "Starting task queue {name} with {} workers",
+                config.config.num_workers
+            );
+            let (tx, rx) = async_channel::bounded(config.channel_size);
+            for n in 0..config.config.num_workers {
+                tracing::debug!("Starting task queue {name} worker {n}");
+                let task_fut = task_fn(rx.clone());
+                queue_tasks.push(tokio::task::spawn(task_fut));
+            }
+            self.queues.insert(name, tx);
+        }
+        let feeder = tokio::task::spawn(self.run(poll_interval));
+        tokio::select! {
+            res = futures::future::select_all(queue_tasks) => {
+                let (res, index, _) = res;
+                if let Err(e) = res {
+                    tracing::error!("Task queue {index} panicked: {e}");
+                    return Err(anyhow::anyhow!("Task queue {index} panicked: {e}"));
```

Comment:
With the display implementation of e we might loose information. Lets try to keep the full error it somewhere for the user.